packages/@glimmer/runtime/lib/scanner.ts

```diff
@@ -81,147 +81,103 @@ function compileStatements(statements: WireFormat.Statement[], meta: Compilation
-  let [, tag] = prelude.pop() as WireFormat.Statements.OpenElement;
-  prelude.push([Ops.ClientSideStatement, ClientSide.Ops.OpenComponentElement, tag]);
-  prelude.push([Ops.ClientSideStatement, ClientSide.Ops.DidCreateElement]);
-
-  let attrsSymbol = symbolTable.symbols.length + 1;
-  symbolTable.symbols.push(ATTRS_BLOCK);
-
-  let statements = prelude
-    .concat([[Ops.Yield, attrsSymbol, EMPTY_ARRAY]])
-    .concat(head)
-    .concat(body)
-    .concat([[Ops.ClientSideStatement, ClientSide.Ops.DidRenderLayout]]);
-
-  return new CompilableTemplate(statements, symbolTable);
-}
+const { Ops } = WireFormat;
 
 export default class Scanner {
   constructor(private block: WireFormat.SerializedTemplateBlock, private env: Environment) {
   }
 
   scanEntryPoint(meta: CompilationMeta): Program {
-    let { block, env } = this;
-
-    let statements: WireFormat.Statement[];
-    if (block.prelude && block.head) {
-      statements = block.prelude.concat(block.head).concat(block.statements);
-    } else {
-      statements = block.statements;
-    }
-
-    return new RawProgram(env, meta, statements, block.symbols, block.hasEval).scan();
+    let { env, block } = this;
+    return new RawProgram(env, meta, block.statements, block.symbols, block.hasEval).scan();
   }
 
   scanBlock(meta: CompilationMeta): Block {
-    let { block, env } = this;
-
-    let statements: WireFormat.Statement[];
-    if (block.prelude && block.head) {
-      statements = block.prelude.concat(block.head).concat(block.statements);
-    } else {
-      statements = block.statements;
-    }
-
-    return new RawInlineBlock(env, meta, statements, EMPTY_ARRAY).scan();
+    let { env, block } = this;
+    return new RawInlineBlock(env, meta, block.statements, EMPTY_ARRAY).scan();
   }
 
-  scanLayout(meta: CompilationMeta, attrs: WireFormat.Statements.Attribute[]): Program {
+  scanLayout(meta: CompilationMeta, attrs: WireFormat.Statements.Attribute[], componentName?: string): Program {
     let { block } = this;
-    let { symbols, hasEval } = block;
-
-    if (!block.prelude || !block.head) {
-      throw new Error(`A layout must have a top-level element`);
-    }
+    let { statements, symbols, hasEval } = block;
 
     let symbolTable = { meta, hasEval, symbols };
-    let { statements: prelude } = scanBlock({ statements: block.prelude, parameters: EMPTY_ARRAY }, meta, this.env);
-    let { statements: head } = scanBlock({ statements: [...attrs, ...block.head], parameters: EMPTY_ARRAY }, meta, this.env);
-    let { statements: body } = scanBlock({ statements: block.statements, parameters: EMPTY_ARRAY }, meta, this.env);
 
-    return layout(prelude, head, body, symbolTable);
+    let newStatements: WireFormat.Statement[] = [];
+
+    let toplevel: string | undefined;
+    let inTopLevel = false;
+
+    for(let i = 0; i < statements.length; i++) {
```

Comment:
These seems like a state machine. I'm ok with leaving it but there might be a better way of expressing this. Perhaps it's something we should come back to.